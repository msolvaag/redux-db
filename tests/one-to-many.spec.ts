import * as ReduxDB from "../dist/cjs";
import * as data from "./data";

const schema = {
    User: {
        username: { pk: true }
    },
    BlogPost: {
        id: { pk: true },
        author: { references: "User", relationName: "posts" }
    },
    Comment: {
        id: { pk: true },
        post: { references: "BlogPost", relationName: "comments" },
        author: { references: "User", relationName: "comments" }
    }
};

export interface BlogPost {
    id: number;
    body: string;
}
export interface Comment {
    id: number;
    post: number;
}

interface User {
    username: string;
    name: string;
}

// Schema models
interface UserRecord extends ReduxDB.TableRecord<User> {
    comments: ReduxDB.TableRecordSet<CommentRecord>;
}

interface BlogPostRecord extends ReduxDB.TableRecord<BlogPost> {
    comments: ReduxDB.TableRecordSet<CommentRecord>;
    author: UserRecord;
}

interface CommentRecord extends ReduxDB.TableRecord<Comment> {
    post: BlogPostRecord;
    author: UserRecord;
}

interface MyTables extends ReduxDB.TableMap {
    BlogPost: ReduxDB.Table<BlogPostRecord>;
    Task: ReduxDB.Table<CommentRecord>;
    User: ReduxDB.Table<UserRecord>;
}

const db = ReduxDB.createDatabase<MyTables>(schema);

const session = (fn: (tables: MyTables) => void, state = {}) => {
    const session = db.createSession(state);
    fn(session.tables);
    return session.commit();
};

describe("one to many", () => {
    const state = session(({ BlogPost }) => BlogPost.insert(data.blogPosts));

    test("state is normalized", () =>
        expect(state).toMatchSnapshot());

    const tables = db.selectTables(state);

    const [originalPost] = data.blogPosts;
    const post = tables.BlogPost.get(originalPost.id);

    test("model", () => expect({
        ...post.value,
        comments: post.comments.all().map(c => ({
            ...c.value,
            author: c.author.value
        })),
        author: post.author.value
    }).toMatchObject(originalPost));
});
