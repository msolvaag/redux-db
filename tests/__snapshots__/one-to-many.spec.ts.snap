// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`one to many state is normalized 1`] = `
Object {
  "BlogPost": Object {
    "byId": Object {
      "1": Object {
        "author": "user1",
        "body": "......",
        "id": 1,
      },
      "2": Object {
        "author": "user2",
        "body": "......",
        "id": 2,
      },
    },
    "ids": Array [
      "1",
      "2",
    ],
    "indexes": Object {
      "author": Object {
        "unique": false,
        "values": Object {
          "user1": Array [
            "1",
          ],
          "user2": Array [
            "2",
          ],
        },
      },
    },
    "meta": Object {},
    "name": "BlogPost",
  },
  "Comment": Object {
    "byId": Object {
      "1": Object {
        "author": "user2",
        "comment": ".....",
        "id": 1,
        "post": "1",
      },
      "2": Object {
        "author": "user3",
        "comment": ".....",
        "id": 2,
        "post": "1",
      },
      "3": Object {
        "author": "user3",
        "comment": ".....",
        "id": 3,
        "post": "2",
      },
      "4": Object {
        "author": "user1",
        "comment": ".....",
        "id": 4,
        "post": "2",
      },
      "5": Object {
        "author": "user3",
        "comment": ".....",
        "id": 5,
        "post": "2",
      },
    },
    "ids": Array [
      "1",
      "2",
      "3",
      "4",
      "5",
    ],
    "indexes": Object {
      "author": Object {
        "unique": false,
        "values": Object {
          "user1": Array [
            "4",
          ],
          "user2": Array [
            "1",
          ],
          "user3": Array [
            "2",
            "3",
            "5",
          ],
        },
      },
      "post": Object {
        "unique": false,
        "values": Object {
          "1": Array [
            "1",
            "2",
          ],
          "2": Array [
            "3",
            "4",
            "5",
          ],
        },
      },
    },
    "meta": Object {},
    "name": "Comment",
  },
  "User": Object {
    "byId": Object {
      "user1": Object {
        "name": "User 1",
        "username": "user1",
      },
      "user2": Object {
        "name": "User 2",
        "username": "user2",
      },
      "user3": Object {
        "name": "User 3",
        "username": "user3",
      },
    },
    "ids": Array [
      "user1",
      "user2",
      "user3",
    ],
    "indexes": Object {},
    "meta": Object {},
    "name": "User",
  },
}
`;
